### PR TITLE
ENG-3767: Land DSR traversal visualizer (initial merge from PoC)

### PR DESCRIPTION
**Pull request by gilluminate**
Ticket [ENG-3767]

### Description Of Changes

**Companion PR:** [fidesplus#3572](https://github.com/ethyca/fidesplus/pull/3572)

Lands the DSR traversal visualizer in `main` for the first time.

**About the diff size.** The PoC originated as a fast-moving, largely AI-generated branch (#8099) that wasn't given a substantive code review before it landed on its source branch. The branch on this PR includes both the cherry-pick of that PoC **and** a polish pass that applies our frontend conventions and addresses a first review pass before merge. The intent is that what you're reviewing is the version of the feature we'd actually want to ship — not the PoC's raw state.

Diffing this PR against the PoC will show substantive differences: string-union types replaced with enums, default exports flipped to named, hex literals swapped for palette tokens, antd v6 deprecations fixed, RTK Query cache key bug fixed, hydration-snap on first paint fixed, CodeQL finding on hand-built routes fixed, and more. Please review the **final state** rather than trying to diff-the-diff. The polish commit history is intact if you want to see what was changed and why.

**Feature summary.** A property-scoped, action-typed preview of how a DSR will fan out through configured integrations. Renders an interactive four-lane graph (identity input → systems queried → manual review gates → not touched) with per-card metadata, dataset breakdowns, and dependency edges. Gated behind the new `dsrTraversalVisualizer` beta flag (dev + test on, production off) and reachable at `/dsr-traversal/[propertyKey]/[[...actionType]]` from the "Request workflows" item in the Privacy requests nav group.

### Code Changes

**Backend** (`src/fides/api/graph/preview/`)
* `builder.py` — `TraversalPreviewBuilder` builds the structured preview, falling back to FK-derived edges when full traversal fails
* `reachability.py` — per-integration reachability classification
* `policy_filter.py` — policy-aware data-category filtering with descendant matching
* `schemas.py` — Pydantic response shapes
* `tests/ops/graph/preview/` — builder + reachability tests with shared `conftest.py` fixtures
* Code review fixes: scale optimization for _static_dataset_detail (O(N×M) → O(1) via indexes), custom category dot-prefix fallback in policy_filter.py, warning surfacing from _capture_traversal, deduplication (graph.identity_keys, _edges_from_counts), modern typing aliases 

**Frontend** (`clients/admin-ui/src/features/dsr-traversal-visualizer/`)
* Page + canvas (ReactFlow-based four-lane layout)
* Three node types (identity root, integration, manual task)
* Two edge types (dependency, gates)
* Two side panels (integration details, manual task details)
* Custom four-lane layout (`layout/compute-lane-layout.ts`, `compute-stages.ts`, `compute-column-count.ts`) — pure functions with unit tests
* RTK Query slice with `serializeQueryArgs` that strips `refresh` so regenerate updates the same cache entry
* Hooks: `useLaneCollapseState` (persists collapse prefs to `localStorage`, exposes a `hydrated` flag to gate first-paint and avoid layout snap), `useNodeSelection`, `useTraversalGraph`
* Route shells in `clients/admin-ui/src/pages/dsr-traversal/`
* Type-safe enums for `ActionType`, `Reachability`, `ActionStatus`, `LaneId`; `REACHABILITY_COLOR` typed against `CUSTOM_TAG_COLOR`

**MSW handlers** (`clients/admin-ui/src/mocks/dsr-traversal/`)
* Self-contained fixture rich enough to exercise all card states (stage-1 + stage-2 reachable, gated by manual review, not-touched) via `npm run dev:mock` without seeding the fidesplus DB

**Flag + nav wiring**
* `dsrTraversalVisualizer` flag in `flags.json` (dev + test on, production off)
* `nav-config.tsx` entry under "Privacy requests → Request workflows", gated by the flag + `requiresPlus`

### Steps to Confirm

The visualizer is gated behind the `dsrTraversalVisualizer` flag (on by default in dev/test, off in production).

1. **Smoke against mock data (no backend setup required):**
   ```
   cd clients/admin-ui && npm run dev:mock
   ```
   Navigate to **Privacy requests → Request workflows** in the sidebar. The MSW handlers seed one property ("Mocked Property") with 5 in-flow integrations across 2 stages, 1 manual review task that gates the legal archive, and 2 not-touched systems.
2. Select the property in the picker → confirm the four-lane layout renders with stage 1 / stage 2 grouping in "Will be queried", the Legal review task in "Gated by manual review", and (with the **Show not touched** toggle on) Legacy ERP + Vendor X in the "Not touched" lane.
3. Click an integration card → side panel opens with system, triggered-by sources, manual-review gate (if any), and dataset/collection/field breakdown.
4. Click **Regenerate** → confirm a single `traversal-preview?refresh=true` network call fires and writes back into the same cache entry (no parallel cache entries).
5. Toggle **Access** / **Erasure** → URL updates and the preview refreshes.
6. Collapse a lane via the floating chevron button → preference persists across reload (`localStorage` key `fides:dsr-traversal:lane-collapse:v1`); on reload the canvas renders with the persisted layout from the first paint (no visible re-layout snap).
7. **Backend smoke (optional, requires fidesplus + seeded property/integrations):** hit `GET /api/v1/plus/properties/{property_id}/traversal-preview?action_type=access&include_unreachable=true` and confirm the response matches the `TraversalPreview` schema; pass `refresh=true` to bypass cache.

### Known Follow-ups (deferred)

From an in-session `/code-review` pass, these were judged worth deferring rather than blocking merge:

* **snake_case wire-format names leak into TS internals** (~24 identifiers across the feature). Normalize at the slice boundary via `transformResponse` in a follow-up — same pattern affects other features in the repo.
* **`compute-stages` BFS depth-tracking deserves a diamond-with-extra-hop regression test.** Practical case appears to be correct; a test would lock it in.
* **Custom category matching semantics**: preview uses taxonomy walk + dot-prefix fallback, real DSR uses bare startswith. Aligned for custom categories but user_provided edge case differs (preview is stricter/more correct).

### Pre-Merge Checklist

* [x] Issue requirements met
* [x] All CI pipelines succeeded
* [x] `CHANGELOG.md` updated
  * [ ] Add a https://github.com/ethyca/fides/labels/db-migration label to the entry if your change includes a DB migration
  * [ ] Add a https://github.com/ethyca/fides/labels/high-risk label to the entry if your change includes a high-risk change (i.e. potential for performance impact or unexpected regression) that should be flagged
  * [ ] Updates unreleased work already in Changelog, no new entry necessary
* UX feedback:
  * [ ] All UX related changes have been reviewed by a designer
  * [x] No UX review needed
* Followup issues:
  * [ ] Followup issues created
  * [ ] No followup issues
* Database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!
  * [ ] No migrations
* Documentation:
  * [ ] Documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] Documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] If there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
  * [ ] No documentation updates required

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[ENG-3767]: https://ethyca.atlassian.net/browse/ENG-3767

### CODE REVIEW

File: clients/admin-ui/src/features/dsr-traversal-visualizer/TraversalVisualizerPage.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        integrations: showNotTouched
+          ? data.integrations
+          : data.integrations.filter(
+              (i) => i.reachability !== Reachability.UNREACHABLE,
+            ),
+      }
+    : undefined;
+
+  const goTo = (key: string, action: ActionType = actionType) => {
+    router.replace(`/dsr-traversal/${key}/${action}`);
```
[the remark this answers is not in the document]
Comment (gilluminate) [reply clients/admin-ui/src/features/dsr-traversal-visualizer/TraversalVisualizerPage.tsx]: Fixed in c291af83f3. Replaced the hand-built template-string path with Next.js's `router.replace({ pathname, query })` form using the existing `DSR_TRAVERSAL_PROPERTY_ACTION_ROUTE` constant — Next.js handles dynamic-segment substitution and URL encoding so untrusted property keys can't smuggle path separators or fragments. CodeQL should clear this on the next scan.

File: src/fides/api/graph/preview/builder.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+                parent = edge.f1.collection_address()
+                if parent != node.address:
+                    deps[node.address].append(parent)
+
+        try:
+            traversal = Traversal(self.graph, self.identity_seed)
+            environment: dict[CollectionAddress, Any] = {
+                ROOT_COLLECTION_ADDRESS: [self.identity_seed]
+            }
+            traversal.traverse(environment, capture)
```
Comment (adamsachs) [inline src/fides/api/graph/preview/builder.py:148]: per claude, is there a good reason we're repeating traversal? the constructor implicitly traverses (for verification, as the docstring notes), and then we explicitly `.traverse` with our `capture` function. we should be able to just do the following and save that initial traversal, right? the explicit traverse call should raise any `UnreachableNodesError` or `UnreachableEdgesError` in the same way the constructor does.
```suggestion
            traversal = Traversal(self.graph, self.identity_seed, skip_verification=True)
            environment: dict[CollectionAddress, Any] = {
                ROOT_COLLECTION_ADDRESS: [self.identity_seed]
            }
            traversal.traverse(environment, capture)
```

File: src/fides/api/graph/preview/schemas.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    data_categories: list[str] = Field(default_factory=list)
+    is_identity: bool = False
+
+
+class CollectionDetail(FidesSchema):
+    name: str
+    skipped: bool = Field(
+        default=False,
+        description="Always False in preview — skip_processing collections "
+        "are excluded during graph construction",
```
Comment (adamsachs) [inline src/fides/api/graph/preview/schemas.py]: nit: if it's always false...do we need it here? :) i.e. do we imagine expanding to somehow include skipped nodes in a preview down the line? or is this just to make things extra-explicit in the API? no

File: src/fides/api/graph/preview/policy_filter.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            return False  # defensive; taxonomy shouldn't cycle
+        visited.add(cursor)
+        cursor = parents.get(cursor)
+    # Category is not in the taxonomy (custom/extended). Fall back to
+    # dot-prefix matching so behavior aligns with the real DSR runner
+    # which uses ``category.startswith(target)``. We add a dot boundary
+    # to avoid false positives like "user_provided" matching target "user".
+    if category not in parents:
+        return any(category == t or category.startswith(t + ".") for t in targets)
+    return False
```
Comment (adamsachs) [inline src/fides/api/graph/preview/policy_filter.py:44]: not a blocker on this PR, but it'd be very nice to avoid duplication of this sort of fragile but important logic throughout the codebase. category filtering, in the context of DSR execution/policies, should be consistent - we definitely risk drift by replicating this sort of matching logic.

i realize it may be easier said than done - but did we at least look to try to leverage a shared util here to ensure we keep this consistent? at a minimum, i think we should put in some comments so that we (1) encourage ourselves to fix the duplicated code (2) make it easier to spot the uses of this logic if/when we ever want to tweak it

**Review by adamsachs — APPROVED** [review]
BE here looks solid to me!

just a few nits/notes along the way. nothing is blocking, but i think these are worth double-checking if we haven't already.

File: src/fides/api/graph/preview/builder.py (hunk 2 above)
Comment (JadeCara) [reply src/fides/api/graph/preview/builder.py:148]: Investigated adding `skip_verification=True` but it turns out the two traversals are **not redundant** — they're coupled. The verification traverse populates `.parents` on `TraversalNode` objects via `add_child()`, and the capture traverse reads those via `incoming_edges()` to record dependency relationships. Skipping verification leaves `.parents` empty, so the capture function records no dependencies and the builder falls back to static classification (wrong results). The double-traverse is the cost of reusing `Traversal`'s internal node-wiring as a side effect. Leaving as-is for now.

File: src/fides/api/graph/preview/schemas.py (hunk 3 above)
Comment (JadeCara) [reply src/fides/api/graph/preview/schemas.py]: Agreed — removed the `skipped` field from `CollectionDetail` (both the Pydantic schema and the TS type). `DatasetGraph` already filters out `skip_processing` collections during construction (graph.py L215-225), so it was always `False`. We can add it back if we ever include those collections in the preview.

File: src/fides/api/graph/preview/policy_filter.py (hunk 4 above)
Comment (JadeCara) [reply src/fides/api/graph/preview/policy_filter.py:44]: Agreed — the preview's `policy_filter.py` does a taxonomy-aware parent-chain walk with dot-boundary guards, but the 5 existing call sites in the DSR runner (`policy.py:182`, `filter_results.py:104,118`, `query_config.py:91`, `graph_task.py:1025`) use bare `startswith` without a dot boundary, which risks false positives (e.g. `"user"` matching `"user_provided"`). Consolidating into a shared util would fix that drift risk and the latent bug. Filed as [ENG-3864](https://ethyca.atlassian.net/browse/ENG-3864).
